src/policy_evaluator.rs

```diff
@@ -186,32 +232,77 @@ impl PolicyEvaluator {
-            Ok(res) => {
-                let pol_val_resp: Result<PolicyValidationResponse> = serde_json::from_slice(&res)
-                    .map_err(|e| anyhow!("cannot deserialize policy validation response: {:?}", e));
-                pol_val_resp
-                    .and_then(|pol_val_resp| {
-                        ValidationResponse::from_policy_validation_response(
-                            uid.to_string(),
-                            &req_obj,
-                            &pol_val_resp,
-                        )
-                    })
-                    .unwrap_or_else(|e| {
+        match &mut self.runtime {
+            Runtime::Wapc(wapc_host) => match wapc_host.call("validate", validate_str.as_bytes()) {
+                Ok(res) => {
+                    let pol_val_resp: Result<PolicyValidationResponse> =
+                        serde_json::from_slice(&res).map_err(|e| {
+                            anyhow!("cannot deserialize policy validation response: {:?}", e)
+                        });
+                    pol_val_resp
+                        .and_then(|pol_val_resp| {
+                            ValidationResponse::from_policy_validation_response(
+                                uid.to_string(),
+                                req_obj,
+                                &pol_val_resp,
+                            )
+                        })
+                        .unwrap_or_else(|e| {
+                            error!(
+                                error = e.to_string().as_str(),
+                                "cannot build validation response from policy result"
+                            );
+                            ValidationResponse::reject_internal_server_error(
+                                uid.to_string(),
+                                e.to_string(),
+                            )
+                        })
+                }
+                Err(e) => {
+                    error!(error = e.to_string().as_str(), "waPC communication error");
+                    ValidationResponse::reject_internal_server_error(uid.to_string(), e.to_string())
+                }
+            },
+            Runtime::Burrego(ref mut burrego) => {
+                let burrego_evaluation = burrego.evaluator.evaluate(
+                    burrego.entrypoint_id,
+                    &burrego.input,
+                    &burrego.data,
+                );
+                match burrego_evaluation {
+                    Ok(evaluation_result) => {
+                        let evaluation_result = serde_json::from_value(
+                            evaluation_result
+                                .get(0)
+                                .unwrap()
+                                .get("result")
+                                .unwrap()
+                                .get("response")
+                                .unwrap()
+                                .clone(),
```

Comment:
This looks hard to understand. Can we do something better? Adding some documentation explaining what we are getting back as a json object would help too.

Now that I recall, is this some temporary code? I'm referring to the conversation we had around how to integrate vanilla OPA responses. I guess this code will be replaced by something that tries to unmarshall the json object into and AdmissionReview object. Am I right?

Reply:
Yes, this is exactly what I was referring to. This is going to change.

Reply:
Great!